File: backend/src/main/java/org/szemeremajax/backend/models/Move.java
package org.szemeremajax.backend.models;

/**
 * Describes a move.
 * @param from The square to move the piece from.
 * @param to The square to move the piece to.
 * @param kind The move's type.
 */
public record Move(int from, int to, MoveKind kind) {
    /**
     * Creates a normal move.
     * @param from The originating square.
     * @param to The landing square.
     * @return The move.
     */
    public static Move createNormal(int from, int to) {
        return new Move(from, to, MoveKind.NORMAL);
    }

    /**
     * Creates a capture move.
     * @param from The originating square.
     * @param to The landing square.
     * @return The move.
     */
    public static Move createCapture(int from, int to) {
        return new Move(from, to, MoveKind.CAPTURE);
    }

    /**
     * Creates a promotion move.
     * @param from The originating square.
     * @param to The landing square.
     * @return The move.
     */
    public static Move createPromotion(int from, int to) {
        return new Move(from, to, MoveKind.PROMOTION);
    }

    /**
     * Creates a promotion-via-capture move.
     * @param from The originating square.
     * @param to The landing square.
     * @return The move.
     */
    public static Move createPromotionWithCapture(int from, int to) {
        return new Move(from, to, MoveKind.PROMOTION_WITH_CAPTURE);
    }
}

File: frontend/src/lib/Move.ts
export type MoveKind = 'NORMAL' | 'CAPTURE' | 'PROMOTION' | 'CAPTURE_WITH_PROMOTION';

export default class Move {
    private readonly _from: number;
    private readonly _to: number;
    private readonly _kind: MoveKind;

    public get from(): number {
        return this._from;
    }

    public get to(): number {
        return this._to;
    }

    public get kind(): MoveKind {
        return this._kind;
    }

    constructor(from: number, to: number, kind: MoveKind) {
        this._from = from;
        this._to = to;
        this._kind = kind;
    }
}